flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/Emitter.java

```diff
@@ -138,19 +145,33 @@ private void output(AsyncResult asyncResult) throws InterruptedException {
 						new Exception("An async function call terminated with an exception. " +
 							"Failing the AsyncWaitOperator.", e));
 				}
+			}
 
-				// remove the peeked element from the async collector buffer so that it is no longer
-				// checkpointed
+			// remove the peeked element from the async collector buffer so that it is no longer
+			// checkpointed
+			try {
 				streamElementQueue.poll();
-
-				// notify the main thread that there is again space left in the async collector
-				// buffer
-				checkpointLock.notifyAll();
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
 			}
+		};
+
+		CompletableFuture<Void> writebackFuture = CompletableFuture.runAsync(processingRequest, mailboxExecutor);
```

Comment:
Logic of synchronizing on `checkpointLock` and notifying it should also be hidden from the user.